File: src/typescript/src/config/mesh-config.ts
/**
 * MeshConfig -- declarative configuration object.
 *
 * Holds the raw configuration dictionary and provides convenience methods
 * for loading from JSON files and accessing configuration sections.
 * Configuration can be built programmatically, loaded from a JSON/YAML
 * file, or constructed by the convenience layer's auto-detection logic.
 */

import * as fs from 'fs';

export class MeshConfig {
  readonly raw: Record<string, unknown>;

  constructor(raw?: Record<string, unknown>) {
    this.raw = raw ?? {};
  }

  /**
   * Load configuration from a JSON file.
   *
   * Since this implementation avoids external dependencies, only JSON
   * files are supported natively. For YAML support, parse externally
   * and use fromDict().
   */
  static fromFile(filePath: string): MeshConfig {
    const content = fs.readFileSync(filePath, 'utf-8');
    const data = JSON.parse(content);
    return new MeshConfig(data);
  }

  /** Alias for fromFile. */
  static fromJson(filePath: string): MeshConfig {
    return MeshConfig.fromFile(filePath);
  }

  /** Create a MeshConfig from a plain object. */
  static fromDict(data: Record<string, unknown>): MeshConfig {
    return new MeshConfig(data);
  }

  // -- Section accessors ---------------------------------------------------

  get providers(): Record<string, unknown> {
    return (this.raw.providers as Record<string, unknown>) ?? {};
  }

  get models(): Record<string, unknown> {
    return (this.raw.models as Record<string, unknown>) ?? {};
  }

  get pools(): Record<string, unknown> {
    return (this.raw.pools as Record<string, unknown>) ?? {};
  }

  get secrets(): Record<string, unknown> {
    return (this.raw.secrets as Record<string, unknown>) ?? {};
  }

  get observability(): Record<string, unknown> {
    return (this.raw.observability as Record<string, unknown>) ?? {};
  }

  get storage(): Record<string, unknown> {
    return (this.raw.storage as Record<string, unknown>) ?? {};
  }

  // -- Utility methods -----------------------------------------------------

  get(key: string, defaultValue?: unknown): unknown {
    return key in this.raw ? this.raw[key] : defaultValue;
  }

  /**
   * Create a new MeshConfig with overrides applied.
   *
   * Performs a shallow merge: top-level keys from overrides replace
   * or extend the corresponding keys in this config.
   */
  merge(overrides: Record<string, unknown>): MeshConfig {
    const merged: Record<string, unknown> = { ...this.raw };
    for (const [key, value] of Object.entries(overrides)) {
      if (
        typeof value === 'object' &&
        value !== null &&
        !Array.isArray(value) &&
        typeof merged[key] === 'object' &&
        merged[key] !== null &&
        !Array.isArray(merged[key])
      ) {
        merged[key] = {
          ...(merged[key] as Record<string, unknown>),
          ...(value as Record<string, unknown>),
        };
      } else {
        merged[key] = value;
      }
    }
    return new MeshConfig(merged);
  }

  /**
   * Run basic validation on the configuration.
   *
   * @returns A list of validation error messages. An empty list means
   *     the configuration is valid.
   */
  validate(): string[] {
    const errors: string[] = [];

    if ('providers' in this.raw && typeof this.raw.providers !== 'object') {
      errors.push("'providers' must be an object");
    }

    if ('models' in this.raw && typeof this.raw.models !== 'object') {
      errors.push("'models' must be an object");
    }

    if ('pools' in this.raw && typeof this.raw.pools !== 'object') {
      errors.push("'pools' must be an object");
    }

    // Check that pool models reference known model IDs
    if ('pools' in this.raw && 'models' in this.raw) {
      const knownModels = new Set(Object.keys(this.models));
      const poolsDef = this.pools;
      for (const [poolId, poolDef] of Object.entries(poolsDef)) {
        const pool = poolDef as Record<string, unknown>;
        const modelList = pool.models as string[] | undefined;
        if (modelList) {
          for (const modelRef of modelList) {
            if (!knownModels.has(modelRef)) {
              errors.push(
                `Pool '${poolId}' references unknown model '${modelRef}'`
              );
            }
          }
        }
      }
    }

    return errors;
  }
}

File: src/python/modelmesh/config/mesh_config.py
"""MeshConfig — declarative configuration object.

Holds the raw configuration dictionary and provides convenience methods
for loading from YAML files and accessing configuration sections.
Configuration can be built programmatically, loaded from a YAML file,
or constructed by the convenience layer's auto-detection logic.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

__all__ = ["MeshConfig"]


@dataclass
class MeshConfig:
    """Declarative configuration for ModelMesh.

    The ``raw`` dict mirrors the YAML configuration structure::

        {
            "secrets": {"store": "modelmesh.env.v1"},
            "providers": { ... },
            "models": { ... },
            "pools": { ... },
            "observability": { ... },
            "storage": { ... },
        }

    Full schema reference: ``docs/SystemConfiguration.md``.

    Attributes:
        raw: The raw configuration dictionary.
    """

    raw: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_yaml(cls, path: str) -> MeshConfig:
        """Load configuration from a YAML file.

        Args:
            path: Path to a YAML configuration file.

        Returns:
            A ``MeshConfig`` populated from the file contents.

        Raises:
            FileNotFoundError: If the file does not exist.
            yaml.YAMLError: If the file is not valid YAML.
        """
        import yaml

        with open(path) as f:
            data = yaml.safe_load(f)
        return cls(raw=data if data is not None else {})

    @classmethod
    def from_file(cls, path: str) -> MeshConfig:
        """Alias for :meth:`from_yaml` for API compatibility."""
        return cls.from_yaml(path)

    @classmethod
    def from_dict(cls, data: dict) -> MeshConfig:
        """Create a MeshConfig from a Python dictionary.

        Args:
            data: Configuration dictionary.

        Returns:
            A ``MeshConfig`` populated from the dictionary.
        """
        return cls(raw=data)

    # -- Section accessors ---------------------------------------------------

    @property
    def providers(self) -> dict[str, Any]:
        """The ``providers`` configuration section."""
        return self.raw.get("providers", {})

    @property
    def models(self) -> dict[str, Any]:
        """The ``models`` configuration section."""
        return self.raw.get("models", {})

    @property
    def pools(self) -> dict[str, Any]:
        """The ``pools`` configuration section."""
        return self.raw.get("pools", {})

    @property
    def secrets(self) -> dict[str, Any]:
        """The ``secrets`` configuration section."""
        return self.raw.get("secrets", {})

    @property
    def observability(self) -> dict[str, Any]:
        """The ``observability`` configuration section."""
        return self.raw.get("observability", {})

    @property
    def storage(self) -> dict[str, Any]:
        """The ``storage`` configuration section."""
        return self.raw.get("storage", {})

    # -- Utility methods -----------------------------------------------------

    def get(self, key: str, default: Any = None) -> Any:
        """Access a top-level configuration key with a default.

        Args:
            key: Top-level key name.
            default: Value to return if key is absent.
        """
        return self.raw.get(key, default)

    def merge(self, overrides: dict[str, Any]) -> MeshConfig:
        """Create a new MeshConfig with overrides applied.

        Performs a shallow merge: top-level keys from *overrides*
        replace or extend the corresponding keys in this config.

        Args:
            overrides: Dictionary of overrides.

        Returns:
            A new ``MeshConfig`` with merged values.
        """
        merged = {**self.raw}
        for key, value in overrides.items():
            if isinstance(value, dict) and isinstance(merged.get(key), dict):
                merged[key] = {**merged[key], **value}
            else:
                merged[key] = value
        return MeshConfig(raw=merged)

    def validate(self) -> list[str]:
        """Run basic validation on the configuration.

        Returns:
            A list of validation error messages. An empty list means
            the configuration is valid.
        """
        errors: list[str] = []

        # Providers section
        if "providers" in self.raw and not isinstance(self.raw["providers"], dict):
            errors.append("'providers' must be a dict")

        # Models section
        if "models" in self.raw and not isinstance(self.raw["models"], dict):
            errors.append("'models' must be a dict")

        # Pools section
        if "pools" in self.raw and not isinstance(self.raw["pools"], dict):
            errors.append("'pools' must be a dict")

        # Check that pool models reference known model IDs
        if "pools" in self.raw and "models" in self.raw:
            known_models = set(self.raw["models"].keys())
            for pool_id, pool_def in self.raw.get("pools", {}).items():
                for model_ref in pool_def.get("models", []):
                    if model_ref not in known_models:
                        errors.append(
                            f"Pool '{pool_id}' references unknown model "
                            f"'{model_ref}'"
                        )

        return errors
